crates/oxc_semantic/tests/util/mod.rs

```diff
@@ -0,0 +1,300 @@
+    data: Result<SymbolId, oxc_diagnostics::Error>,
+}
+
+impl<'a> SymbolTester<'a> {
+    #[allow(dead_code)]
+    pub(super) fn new_at_root(
+        parent: &'a SemanticTester,
+        semantic: Semantic<'a>,
+        target: &str,
+    ) -> Self {
+        let decl =
+            semantic.scopes().get_binding(semantic.scopes().root_scope_id(), &Atom::from(target));
+        let data = decl.map_or_else(|| Err(miette!("Could not find declaration for {target}")), Ok);
+
+        SymbolTester { parent, semantic, target: target.to_string(), data }
+    }
+
+    pub(super) fn new_unique(
+        parent: &'a SemanticTester,
+        semantic: Semantic<'a>,
+        target: &str,
+    ) -> Self {
+        let symbols_with_target_name: Vec<_> =
+            semantic.scopes().iter_bindings().filter(|(_, _, name)| name == &target).collect();
+        let data = match symbols_with_target_name.len() {
+            0 => Err(miette!("Could not find declaration for {target}")),
+            1 => Ok(symbols_with_target_name.iter().map(|(_, symbol_id, _)| *symbol_id).next().unwrap()),
+            n if n > 1 => Err(miette!("Couldn't uniquely resolve symbol id for target {target}; {n} symbols with that name are declared in the source.")),
+            _ => unreachable!()
+        };
+
+        SymbolTester { parent, semantic, target: target.to_string(), data }
+    }
+
+    /// Checks if the resolved symbol contains all flags in `flags`, using [`SymbolFlags::contains()`]
+    pub fn contains_flags(mut self, flags: SymbolFlags) -> Self {
+        self.data = match self.data {
+            Ok(symbol_id) => {
+                let found_flags = self.semantic.symbols().get_flag(symbol_id);
+                if found_flags.contains(flags) {
+                    Ok(symbol_id)
+                } else {
+                    // let decl_span = self.semantic.symbol_declaration(symbol_id).span();
+                    Err(miette!(
+                        "Expected {} to contain flags {:?}, but it had {:?}",
+                        self.target,
+                        flags,
+                        found_flags
+                    ))
+                }
+            }
+            err => err,
+        };
+        self
+    }
+
+    pub fn intersects_flags(mut self, flags: SymbolFlags) -> Self {
+        self.data = match self.data {
+            Ok(symbol_id) => {
+                let found_flags = self.semantic.symbols().get_flag(symbol_id);
+                if found_flags.intersects(flags) {
+                    Ok(symbol_id)
+                } else {
+                    Err(miette!(
+                        "Expected {} to intersect with flags {:?}, but it had {:?}",
+                        self.target,
+                        flags,
+                        found_flags
+                    ))
+                }
+            }
+            err => err,
+        };
+        self
+    }
+
+    pub fn has_number_of_reads(self, ref_count: usize) -> Self {
```

Comment:
nitpick on the API: what about `assert_xxx` and use `&self`? Chaining these methods calls are equivalent of calling them individually. Calling them individually has the advance of writing test cases where if statements are required. e.g.

```
let tester = Tester::new();
if is_condition {
   tester.assert_xxx();
}
```

Reply:
I personally like the chaining, it builds a sentence. I agree with your first point, how about something like `should_have_read_count`? To me, this sounds more like a predicate than `assert_xxx`

Reply:
Ok, `should_have_read_count` or `has_number_of_reads` are both fine.